src/vos/vos_common.c

```diff
@@ -121,6 +122,106 @@ vos_bio_addr_free(struct vos_pool *pool, bio_addr_t *addr, daos_size_t nob)
 	return rc;
 }
 
+static int
+vos_tx_publish(struct dtx_handle *dth, int err)
+{
+	struct vos_container	*cont = vos_hdl2cont(dth->dth_coh);
+	struct dtx_rsrvd_uint	*dru;
+	int			 rc;
+	int			 i;
+
+again:
+	for (i = 0, rc = 0;
+	     i < dth->dth_modification_cnt && (rc == 0 || err != 0); i++) {
+		dru = &dth->dth_rsrvds[i];
+		if (dru->dru_scm.drs_ptr != NULL) {
+			rc = vos_publish_scm(cont, &dru->dru_scm, err == 0);
+
+			/* FIXME: Currently, vos_publish_blocks() will release
+			 *	  reserved information in 'dru_nvme_list' from
+			 *	  DRAM. So if vos_publish_blocks() failed, the
+			 *	  former published ones cannot be rollback. It
+			 *	  will be handled later.
+			 */
+			if (rc == 0 || err != 0)
+				rc = vos_publish_blocks(cont, &dru->dru_nvme,
+							err == 0,
+							VOS_IOS_GENERIC);
+			if (err != 0)
+				D_FREE(dru->dru_scm.drs_ptr);
+		}
+	}
+
+	/* Some publish failed, cancel all. */
```

Comment:
This won't work for this moment. If published is called, the actv array for scm reserve and the reserved list for nvme reserve will be freed.

Reply:
Yes, currently, we only can handle scm reservation publish/cancel, we need more work for nvme case. But that is out of this patch scope.